development/webpack/utils/helpers.ts

```diff
@@ -0,0 +1,243 @@
+import { readdirSync } from 'node:fs';
+import { parse, join, relative, sep } from 'node:path';
+import type { Chunk, EntryObject, Stats } from 'webpack';
+import type TerserPluginType from 'terser-webpack-plugin';
+
+export type Manifest = chrome.runtime.Manifest;
+export type ManifestV2 = chrome.runtime.ManifestV2;
+export type ManifestV3 = chrome.runtime.ManifestV3;
+
+// HMR (Hot Module Reloading) can't be used until all circular dependencies in
+// the codebase are removed
+// See: https://github.com/MetaMask/metamask-extension/issues/22450
+// TODO: remove this variable when HMR is ready. The env var is for tests and
+// must also be removed everywhere.
+export const __HMR_READY__ = Boolean(process.env.__HMR_READY__) || false;
+
+/**
+ * Target browsers
+ */
+export const Browsers = ['brave', 'chrome', 'firefox', 'opera'] as const;
+export type Browser = (typeof Browsers)[number];
+
+const slash = `(?:\\${sep})?`;
+/**
+ * Regular expression to match files in any `node_modules` directory
+ * Uses a platform-specific path separator: `/` on Unix-like systems and `\` on
+ * Windows.
+ */
+export const NODE_MODULES_RE = new RegExp(`${slash}node_modules${slash}`, 'u');
+
+/**
+ * No Operation. A function that does nothing and returns nothing.
+ *
+ * @returns `undefined`
+ */
+export const noop = () => undefined;
+
+/**
+ *
+ * @returns Returns the current version of MetaMask as specified in package.json
+ */
+export const getMetaMaskVersion = (): string => {
+  const { version } = require('../../../package.json');
+  return version as string;
+};
+
+/**
+ * Collects all entry files for use with webpack.
+ *
+ * TODO: move this logic into the ManifestPlugin
```

Comment:
This `TODO:` is here because it will add a bit more complexity to make this part of a webpack plugin. Having it do so much heavy lifting before webpack compilation isn't the "webpack way". The biggest reason for moving it to a plugin is that in `watch` mode we won't be able to detect new `entries` (added to the manifest file or HTML files added to the app directory).

Another reason is that it won't be part of the webpack cache (but the performance gain is negligible anyway).